### PR TITLE
[WIP] Upgrade to GLFW 3.2.1

### PR DESCRIPTION
**Pull request by cloudhead**
- [x] Builds on Linux
- [x] Builds on macOS
- [x] Builds on Windows
- [x] Existing tests pass
- [x] New bindings implemented
- [x] New tests pass

Since a lot is also missing from 3.1 (see #19), I will include the missing changes in this branch:
- [x] Added GLFWcursor custom system cursor handle
- [x] Added glfwCreateCursor, glfwCreateStandardCursor, glfwDestroyCursor and glfwSetCursor for managing system cursor images
- [x] Added GLFWimage struct for passing 32-bit RGBA images
- [x] Added monitor and adapter identifier access to native API
- [x] Added glfwSetDropCallback and GLFWdropfun for receiving dropped files
- [x] Added glfwPostEmptyEvent for allowing secondary threads to cause glfwWaitEvents to return
- [x] Added glfwSetCharModsCallback for receiving character events with modifiers
- [x] Added glfwGetWindowFrameSize for retrieving the size of the frame around the client area of a window
- [x] Added GLFW_AUTO_ICONIFY for controlling whether full screen windows automatically iconify (and restore the previous video mode) on focus loss
- [x] Added GLFW_DONT_CARE for indicating that any value is acceptable
- [x] Added GLFW_DOUBLEBUFFER for controlling whether to use double buffering
- [x] Added GLFW_CONTEXT_RELEASE_BEHAVIOR and values GLFW_ANY_RELEASE_BEHAVIOR, GLFW_RELEASE_BEHAVIOR_FLUSH and GLFW_RELEASE_BEHAVIOR_NONE for GL_KHR_context_flush_control support
- [x] Added GLFW_FLOATING for creating always-on-top windowed mode windows
- [x] Added GLFW_FOCUSED window hint for controlling initial input focus

The following will be implemented from GLFW 3.2:

- [x] Added glfwVulkanSupported, glfwGetRequiredInstanceExtensions, glfwGetInstanceProcAddress, glfwGetPhysicalDevicePresentationSupport and glfwCreateWindowSurface for platform independent Vulkan support
- [x] Added glfwSetWindowMonitor for switching between windowed and full screen modes and updating the monitor and desired video mode of full screen windows
- [x] Added glfwMaximizeWindow and GLFW_MAXIMIZED for window maximization
- [x] Added glfwFocusWindow for giving windows input focus
- [x] Added glfwSetWindowSizeLimits and glfwSetWindowAspectRatio for setting absolute and relative window size limits
- [x] Added glfwGetKeyName for querying the layout-specific name of printable keys
- [x] Added glfwWaitEventsTimeout for waiting for events for a set amount of time
- [x] Added glfwSetWindowIcon for setting the icon of a window
- [x] Added glfwGetTimerValue and glfwGetTimerFrequency for raw timer access
- [x] Added glfwSetJoystickCallback and GLFWjoystickfun for joystick connection and disconnection events
- [x] Added GLFW_NO_API for creating window without contexts
- [x] Added GLFW_INCLUDE_VULKAN for including the Vulkan header
- [x] Added GLFW_CONTEXT_CREATION_API, GLFW_NATIVE_CONTEXT_API and GLFW_EGL_CONTEXT_API for run-time context creation API selection
- [x] Added GLFW_CONTEXT_NO_ERROR context hint for GL_KHR_no_error support

### CODE REVIEW

**Comment by YellowOnion** [comment]
Can I help in anyway? I'm on Windows (I guess I can test if it builds)

**Comment by cloudhead** [comment]
@YellowOnion yeah, as I don't have windows around it would be helpful to test the build, and once I'm finished with the tests, to run the tests on your machine.

File: Test.hs [hunk 1]
```diff
@@ -419,6 +460,10 @@ test_glfwWaitEvents :: IO ()
 test_glfwWaitEvents =
     c'glfwWaitEvents
 
+test_glfwWaitEventsTimeout :: IO ()
+test_glfwWaitEventsTimeout =
+    c'glfwWaitEventsTimeout 1.0
```
Comment (OlivierSohn) [inline Test.hs]: I think it will make the test one second longer, are you ok with changing this value to 0.001 ?

**Comment by OlivierSohn** [comment]
Once #54 is merged in this 3.2.1 branch, we can check the "Builds on macOS"

**Comment by Mokosha** [comment]
I've fixed the branch conflicts and windows build issues in #58

**Comment by Mokosha** [comment]
This branch is ready to be merged into master, it just needs to be tested on Linux.

**Comment by Mokosha** [comment]
Commit 17356f1 fixes #33.

**Comment by achirkin** [comment]
Don't know if it helps, but here is what happens if I run it on my machine.
`Ubuntu 17.10 | 4.13.0-17-generic x86_64 | GeForce GTX 960 | nvidia-390 `
```
bindings-GLFW-3.2.1.0: test (suite: main)

Initialization and version information:
  glfwGetVersion: [OK]
  glfwGetVersionString: [OK]
Monitor handling:
  glfwGetMonitors: [OK]
  glfwGetPrimaryMonitor: [OK]
  glfwGetMonitorPos: [OK]
  glfwGetMonitorPhysicalSize: [OK]
  glfwGetMonitorName: [OK]
  glfwGetVideoModes: [OK]
  glfwGetVideoMode: [OK]
  glfwGetGammaRamp: [OK]
Window handling:
  glfwDefaultWindowHints: [OK]
  glfwGetWindowAttrib: [OK]
  window close flag: [OK]
  glfwSetWindowTitle: [OK]
  window pos: [OK]
  window size: [OK]
  glfwGetWindowFrameSize: [OK]
  glfwGetFramebufferSize: [OK]
  iconification: [Failed]
expected: 1
 but got: 0
  glfwGetWindowMonitor: [OK]
  glfwSetWindowMonitor: [OK]
  glfwSetWindowIcon: [OK]
  glfwMaximizeWindow: [Failed]
expected: 1
 but got: 0
  glfwSetWindowSizeLimits: [OK]
  glfwSetWindowAspectRatio: [OK]
  glfwFocusWindow: [OK]
  cursor pos: [Failed]
expected: 50.0
 but got: 3768.0
  glfwPollEvents: [OK]
  glfwWaitEvents: [OK]
  glfwWaitEventsTimeout: [OK]
Input handling:
  glfwJoystickPresent: [OK]
  glfwGetJoystickAxes: [OK]
  glfwGetJoystickButtons: [OK]
  glfwGetJoystickName: [OK]
  glfwGetKeyName: [OK]
Time:
  glfwGetTime: [OK]
  glfwSetTime: [OK]
Timers:
  glfwGetTimerValue: [OK]
  glfwSetTimerFrequency: [OK]
Context:
  glfwGetCurrentContext: [OK]
  glfwSwapBuffers: [OK]
  glfwSwapInterval: [OK]
  glfwExtensionSupported: [OK]
Clipboard:
  clipboard: [OK]bindings-GLFW-3.2.1.0: test (suite: main)

Initialization and version information:
  glfwGetVersion: [OK]
  glfwGetVersionString: [OK]
Monitor handling:
  glfwGetMonitors: [OK]
  glfwGetPrimaryMonitor: [OK]
  glfwGetMonitorPos: [OK]
  glfwGetMonitorPhysicalSize: [OK]
  glfwGetMonitorName: [OK]
  glfwGetVideoModes: [OK]
  glfwGetVideoMode: [OK]
  glfwGetGammaRamp: [OK]
Window handling:
  glfwDefaultWindowHints: [OK]
  glfwGetWindowAttrib: [OK]
  window close flag: [OK]
  glfwSetWindowTitle: [OK]
  window pos: [OK]
  window size: [OK]
  glfwGetWindowFrameSize: [OK]
  glfwGetFramebufferSize: [OK]
  iconification: [Failed]
expected: 1
 but got: 0
  glfwGetWindowMonitor: [OK]
  glfwSetWindowMonitor: [OK]
  glfwSetWindowIcon: [OK]
  glfwMaximizeWindow: [Failed]
expected: 1
 but got: 0
  glfwSetWindowSizeLimits: [OK]
  glfwSetWindowAspectRatio: [OK]
  glfwFocusWindow: [OK]
  cursor pos: [Failed]
expected: 50.0
 but got: 3768.0
  glfwPollEvents: [OK]
  glfwWaitEvents: [OK]
  glfwWaitEventsTimeout: [OK]
Input handling:
  glfwJoystickPresent: [OK]
  glfwGetJoystickAxes: [OK]
  glfwGetJoystickButtons: [OK]
  glfwGetJoystickName: [OK]
  glfwGetKeyName: [OK]
Time:
  glfwGetTime: [OK]
  glfwSetTime: [OK]
Timers:
  glfwGetTimerValue: [OK]
  glfwSetTimerFrequency: [OK]
Context:
  glfwGetCurrentContext: [OK]
  glfwSwapBuffers: [OK]
  glfwSwapInterval: [OK]
  glfwExtensionSupported: [OK]
Clipboard:
  clipboard: [OK]
Vulkan:
  glfwVulkanSupported: [OK]
  glfwGetRequiredInstanceExtensions: [OK]
  glfwGetInstanceProcAddress: [OK]
### error: 65542 "X11: Vulkan instance missing VK_KHR_xcb_surface extension"
  glfwGetPhysicalDevicePresentationSupport: [OK]
### error: 65542 "X11: Vulkan instance missing VK_KHR_xcb_surface extension"
  glfwCreateWindowSurface: [OK]

         Test Cases   Total       
 Passed  46           46          
 Failed  3            3           
 Total   49           49          


Vulkan:
  glfwVulkanSupported: [OK]
  glfwGetRequiredInstanceExtensions: [OK]
  glfwGetInstanceProcAddress: [OK]
### error: 65542 "X11: Vulkan instance missing VK_KHR_xcb_surface extension"
  glfwGetPhysicalDevicePresentationSupport: [OK]
### error: 65542 "X11: Vulkan instance missing VK_KHR_xcb_surface extension"
  glfwCreateWindowSurface: [OK]

         Test Cases   Total       
 Passed  46           46          
 Failed  3            3           
 Total   49           49          


```

**Comment by OlivierSohn** [comment]
#59 may fix one of the failing tests (cursor pos) on Linux. It does so on OSX.

**Comment by Mokosha** [comment]
@achirkin Thanks, that does help. Might be OK that the iconify and maximize fns don't pass on linux -- it might be a GLFW bug. Will investigate.

@OlivierSohn I'm going to merge your PR because in all likelihood it doesn't hurt. Thanks!

**Comment by OlivierSohn** [comment]
@Mokosha I'm on 10.13.3, I think that's the reason we have different behaviours.

**Comment by cloudhead** [comment]
Hey Mokosha, thanks for your work! On linux I get this error now:

```
bindings-GLFW-3.2.1.0: test (suite: main)

X Error of failed request:  BadAtom (invalid Atom parameter)
  Major opcode of failed request:  20 (X_GetProperty)
  Atom id in failed request:  0x0
  Serial number of failed request:  192
  Current serial number in output stream:  192

bindings-GLFW-3.2.1.0: Test suite main failed
Test suite failure for package bindings-GLFW-3.2.1.0
    main:  exited with: ExitFailure 1
Logs printed to console
```
```
  glfwGetWindowFrameSize: [Failed]
Window has no frame width up top!
  iconification: [Failed]
expected: 1
 but got: 0
```

**Comment by achirkin** [comment]
@Mokosha, on my machine, all but one tests are working:
```
...
Window handling:
  ...: [OK]
  cursor pos: [Failed]
...
         Test Cases   Total       
 Passed  48           48          
 Failed  1            1           
 Total   49           49
```
I also tried to investigate this cursor pos failing issue. The problem is `c'glfwSetCursorPos` simply does not change the cursor position for me. Waiting for `WindowFocusCallback` before setting cursor position does not help either. Some people in the internet say that this is even not a problem of GLFW, but rather a configuration of X server.

**Comment by bsl** [comment]
@achirkin Can I ask what window manager you're in? I have a very dim memory of not being able to make that call work, probably because I was in a tiling window manager like Awesome or i3.

**Comment by achirkin** [comment]
@bsl Sorry, forgot to add. I have a standard Ubuntu GNOME 17.10 setup; `GNOME Shell 3.26.2`. Runs on X with proprietary `nvidia-390` drivers.

**Comment by Mokosha** [comment]
Thanks, everyone. I also think that the different window managers are causing different things to fail for different people. If the test passes on any one linux machine, then it's likely OK, and whatever underlying difficulty exists is likely a `GLFW` problem, not a `bindings-GLFW` problem.

Unless there are any objections, I'll go ahead and merge/release this branch as 3.2.1.0 tonight.

**Pull request merged**